### PR TITLE
[SPARK-50092][SQL] Fix PostgreSQL connector behaviour for multidimensional arrays

### PR DESCRIPTION
**Pull request by PetarVasiljevic-DB**
### What changes were proposed in this pull request?
There is a bug introduced in this PR https://github.com/apache/spark/pull/46006. This PR fixed the behaviour for PostgreSQL connector for multidimensional arrays since we have mapped all arrays to 1D arrays.

This PR has introduced a bug for one case. Following scenario is broken:

- User has a table t1 on Postgres and does CTAS command to create table t2 with same data.
PR 46006 is resolving the dimensionality of column by reading the metadata from pg_attribute table and attndims column.
- This query returns correct dimensionality for table t1, but for table t2 that is created via CTAS it returns 0 always. This leads to all of the arrays being mapped to 0-D array which is the type itself (for example int). This is a bug on Postgres side.
- As a solution, we can query array_ndims function on given column that will return the dimension of the column. It works for CTAS-like-created tables too. We can get the result of this function on first row of table. 

This change is doing additional query to PG table to find the dimension of array column instead of querying metadata table as before. It might be more expensive but we are sending LIMIT 1 in query.

Also, there is one caveat. In PG, there is no dimensionality of array as all the arrays are 1D arrays (https://www.postgresql.org/docs/current/arrays.html#ARRAYS-DECLARATION). Therefore, if there is table with 2D array column, it is totally fine from PG side to insert 1D or 3D array in this column. This makes the read path on Spark problematic since we will get the dimension of array, for example 1 if the first record is 1D array, and then we will try to read 3D array later on which will fail. Vice versa also, getting dimension 3 and reading 1D array later on. The change that I propose is fine with this scenario since it already doesn't work in Spark. What my change implies is that user can get different error message depending on the dimensionality of first record in table (namely, for one table they can get the error message that the expected type is ARRAY<ARRAY<INT>> and for the other that it is ARRAY\<INT\>.

### Why are the changes needed?
Bug fix.

### Does this PR introduce _any_ user-facing change?
No. It just fixes one case that doesn't work currently.

### How was this patch tested?
New tests.

### CODE REVIEW

**Comment by PetarVasiljevic-DB** [comment]
@yaooqinn as an author of https://github.com/apache/spark/pull/46006, I would like to get your review here :).

File: connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/PostgresIntegrationSuite.scala [hunk 1]
```diff
@@ -23,10 +23,10 @@ import org.apache.spark.{SparkConf, SparkSQLException}
 import org.apache.spark.sql.AnalysisException
 import org.apache.spark.sql.catalyst.analysis.TableAlreadyExistsException
 import org.apache.spark.sql.execution.datasources.v2.jdbc.JDBCTableCatalog
+import org.apache.spark.sql.internal.SQLConf
 import org.apache.spark.sql.jdbc.DatabaseOnDocker
 import org.apache.spark.sql.types._
 import org.apache.spark.tags.DockerTest
-
```
Comment (uros-db) [inline connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/PostgresIntegrationSuite.scala:29]: undo this change please

**Review by MaxGekk — COMMENTED** [review]
> @yaooqinn as an author of https://github.com/apache/spark/pull/46006, I would like to get your review here :).

also cc @dongjoon-hyun as you reviewed the original PR.

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala [hunk 2]
```diff
@@ -382,18 +382,21 @@ private case class PostgresDialect()
       case Types.ARRAY =>
         val tableName = rsmd.getTableName(columnIdx)
         val columnName = rsmd.getColumnName(columnIdx)
-        val query =
+        val query = if (conf.postgresProperReadingOfArrayDimensionality) {
+             s"SELECT array_ndims($columnName)) FROM $tableName LIMIT 1"
+        } else {
           s"""
              |SELECT pg_attribute.attndims
              |FROM pg_attribute
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala:390]: do you mean there is a bug in pgsql information schema which returns wrong `attndims`?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala:390]: interesting, let's leave a code comment here to explain it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 3]
```diff
@@ -4421,6 +4421,16 @@ object SQLConf {
       .booleanConf
       .createWithDefault(false)
 
+  val POSTGRES_PROPER_READING_OF_ARRAY_DIMENSIONALITY =
+    buildConf("spark.sql.legacy.postgres.properReadingOfArrayDimensionality")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:4425]: usually legacy config true means use the legacy behavior.  how about `spark.sql.legacy.postgres.getArrayDimFrom PgAttribute `.

And usually we enable bug fix by default.

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala (hunk 2 above)
Comment (PetarVasiljevic-DB) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala:396]: Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 3 above)
Comment (PetarVasiljevic-DB) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Thanks for suggestion. Changed the code now.

File: connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/PostgresIntegrationSuite.scala [hunk 4]
```diff
@@ -125,43 +126,58 @@ class PostgresIntegrationSuite extends DockerJDBCIntegrationV2Suite with V2JDBCT
...
-      ("array_int", "\"ARRAY<INT>\""),
-      ("array_bigint", "\"ARRAY<BIGINT>\""),
-      ("array_smallint", "\"ARRAY<SMALLINT>\""),
-      ("array_boolean", "\"ARRAY<BOOLEAN>\""),
-      ("array_float", "\"ARRAY<FLOAT>\""),
-      ("array_double", "\"ARRAY<DOUBLE>\""),
-      ("array_timestamp", "\"ARRAY<TIMESTAMP>\""),
-      ("array_timestamptz", "\"ARRAY<TIMESTAMP>\"")
-    )
+    withSQLConf(SQLConf.POSTGRES_GET_ARRAY_DIM_FROM_PG_ATTRIBUTE_METADATA_TABLE.key -> "true") {
```
Comment (PetarVasiljevic-DB) [inline connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/PostgresIntegrationSuite.scala:144]: Have to change the config here because it wouldn't crash with new code since arrays will be read as 2D and not as 1D as it in old path.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 5]
```diff
@@ -4421,6 +4421,16 @@ object SQLConf {
       .booleanConf
       .createWithDefault(false)
 
+  val POSTGRES_GET_ARRAY_DIM_FROM_PG_ATTRIBUTE_METADATA_TABLE =
```
Comment (yaooqinn) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:4424]: We don't need this new parameter if pg's array_ndims satisfies all cases, we can completely remove the branch that visits the information_schema.

Comment (PetarVasiljevic-DB) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:4424]: I agree. I wanted to have the config just in case something goes wrong so we can fallback to previous behaviour.

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:4424]: There is no 'previous/legacy behavior', this feature doesn't belong to any final official releases yet. There is nothing to restore. So, adding a legacy configuration is overkilling

Comment (PetarVasiljevic-DB) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Thanks. Done.

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala [hunk 6]
```diff
@@ -382,18 +382,11 @@ private case class PostgresDialect()
...
         val columnName = rsmd.getColumnName(columnIdx)
-        val query =
-          s"""
-             |SELECT pg_attribute.attndims
-             |FROM pg_attribute
-             |  JOIN pg_class ON pg_attribute.attrelid = pg_class.oid
-             |  JOIN pg_namespace ON pg_class.relnamespace = pg_namespace.oid
-             |WHERE pg_class.relname = '$tableName' and pg_attribute.attname = '$columnName'
-             |""".stripMargin
+        val query = s"SELECT array_ndims($columnName) FROM $tableName LIMIT 1"
```
Comment (milastdbx) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala:385]: Lets add comment

```
// Spark does not support different array dimensionality per row. With this in mind, get dimensionality of any row.
// This might fail down the road, as Postgres array column might contain different dimensionality.
```

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala [hunk 7]
```diff
@@ -382,18 +382,11 @@ private case class PostgresDialect()
...
         try {
           Using.resource(conn.createStatement()) { stmt =>
             Using.resource(stmt.executeQuery(query)) { rs =>
-              if (rs.next()) metadata.putLong("arrayDimension", rs.getLong(1))
+              if (rs.next()) metadata.putLong("arrayDimension", Math.max(rs.getLong(1), 1))
```
Comment (milastdbx) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala:389]: Add comments
```
// Depending on how table is created dimensionality of array might not be present in metadata.
// In these cases above pg metadata query returns zero for arrayDimension. We default to 1 in this case.
```

**Review by milastdbx — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 5 above)
Comment (PetarVasiljevic-DB) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: @yaooqinn may I get review again?

File: connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/PostgresIntegrationSuite.scala [hunk 8]
```diff
@@ -302,4 +317,22 @@ class PostgresIntegrationSuite extends DockerJDBCIntegrationV2Suite with V2JDBCT
...
+
+  test("Test reading 2d array from table created via CTAS command") {
+    val dfWithNewBehaviour = sql(s"SELECT * FROM $catalogName.array_of_array_of_int")
+    val CTASdfWithNewBehaviour = sql(s"SELECT * FROM $catalogName.ctas_array_of_array_of_int")
+
+    checkAnswer(CTASdfWithNewBehaviour, dfWithNewBehaviour.collect())
+  }
+
+  test("Test reading multiple dimension array from table created via CTAS command") {
+    val exception = intercept[org.apache.spark.SparkSQLException] {
```
Comment (yaooqinn) [inline connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/PostgresIntegrationSuite.scala:329]: Can we use `checkError` here? BTW, we just added this test in passing, not because of a behavioural change I guess?

**Comment by yaooqinn** [comment]
Does this approach still work when the resultset given by limit 1 is empty?

**Comment by cloud-fan** [comment]
> Does this approach still work when the resultset given by limit 1 is empty?

This is a good point. Shall we keep the information schema query as a fallback when the table is empty?

**Comment by PetarVasiljevic-DB** [comment]
> Does this approach still work when the resultset given by limit 1 is empty?
This is actually fine as we won't set arrayDimension in metadata. Later on, when reading we are doing something like 
```
val dim = if (metadata.contains("arrayDimension")) {
  metadata.getLong("arrayDimension").toInt
} else {
  1
}```

so we are always going to fallback to 1. This is also what happened so far, if metadata for some reason doesn't contain info about dimensionality we fallback to 1. I can write tests though to make sure this works.

**Comment by yaooqinn** [comment]
If we execute a CTAS on the Spark side against an empty Postgres table with multidimensional array columns, the new approach now defaults the dimension value to 1, which might cause subsequent data pipelines to fail. While the information schema approach was initially impacted by an 'upstream' bug, it now appears that we have introduced a bug of our own.

File: connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/PostgresIntegrationSuite.scala (hunk 8 above)
Comment (PetarVasiljevic-DB) [reply connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/PostgresIntegrationSuite.scala]: Done

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala [hunk 9]
```diff
@@ -382,18 +382,45 @@ private case class PostgresDialect()
...
+         different dimensions per row for arrays.
+         */
+        val query = s"SELECT array_ndims($columnName) FROM $tableName LIMIT 1"
+        var arrayDimensionalityResolveNeedsFallback = true
+
         try {
           Using.resource(conn.createStatement()) { stmt =>
             Using.resource(stmt.executeQuery(query)) { rs =>
-              if (rs.next()) metadata.putLong("arrayDimension", rs.getLong(1))
+              if (rs.next()) {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala:397]: is there a `rs.hasNext` function?

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala [hunk 10]
```diff
@@ -382,18 +382,45 @@ private case class PostgresDialect()
...
+                metadata.putLong("arrayDimension", rs.getLong(1))
+                arrayDimensionalityResolveNeedsFallback = false
+              }
+            }
+          }
+
+          if (arrayDimensionalityResolveNeedsFallback) {
```
Comment (yaooqinn) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala:404]: we can simply move this into the else-block of the first ` if (rs.next())` without creating `arrayDimensionalityResolveNeedsFallback `

Comment (PetarVasiljevic-DB) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala]: I thought it would be cleaner with additional nesting of `Using.resource...`. so I introduced `arrayDimensionalityResolveNeedsFallback`. I have refactored now.

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala (hunk 9 above)
Comment (PetarVasiljevic-DB) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala:396]: Well rs.next() returns false if the resultset is empty so we can just reuse it. I have updated the PR.

**Comment by yaooqinn** [comment]
One more question about array_ndims, does it return a constant for a specific column, in other words, is it possible that Postgres does not update system meta because the dimension value from CTAS is not deterministic?

**Comment by PetarVasiljevic-DB** [comment]
array_ndims will return the dimension of the value in specific row. In Postgres it is possible to insert different dimensionality array into the column with predefined array dimension. Therefore, in such case, different rows will give different value of array_ndims. 

For CTAS tables, if there are various dimensionality values than I would say it is kinda fine that Postgres doesn't update meta table since it's not deterministic, but if all the values have same dimensionality (and same as the dimension read from meta for old table) then meta table should have that information but it actually doesn't.

LMK if I haven't answered to your question.

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala [hunk 11]
```diff
@@ -382,18 +382,42 @@ private case class PostgresDialect()
...
-          s"""
-             |SELECT pg_attribute.attndims
-             |FROM pg_attribute
-             |  JOIN pg_class ON pg_attribute.attrelid = pg_class.oid
-             |  JOIN pg_namespace ON pg_class.relnamespace = pg_namespace.oid
-             |WHERE pg_class.relname = '$tableName' and pg_attribute.attname = '$columnName'
-             |""".stripMargin
+
+        /*
+         Spark does not support different dimensionality per row, therefore we retrieve the
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala:387]: If the pgsql table was not created and inserted by Spark, it's possible that values in different rows have different dimensionality, right? In that case, is it really ok to get dimensionality from the first row?

Comment (PetarVasiljevic-DB) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala:387]: I would say it is ok as the read will fail anyways. Previously we were reading from metadata, and whatever dimension we get there, if in table we have a row with different dimension read will fail.

**Comment by yaooqinn** [comment]
I conducted a test myself, and I'm now pondering its validity.

image

**Comment by PetarVasiljevic-DB** [comment]
What is your concern here?

**Comment by PetarVasiljevic-DB** [comment]
@yaooqinn are you questioning the reasoning behind metadata not being correct for CTAS created tables or I misunderstood your question?

**Comment by yaooqinn** [comment]
Hi @PetarVasiljevic-DB,

The experiment I conducted shows that executing array_ndim on the same column can result in different values across different rows.

**Comment by PetarVasiljevic-DB** [comment]
@yaooqinn yes, this is is purely dependant on the value of the row so it is possible we get different values. I have mentioned in PR description that we can have 2D array and we are allowed to insert 1D or 3D elements. Therefore we can expect returned dimension to be 1 or 3 depending on the first row fetched, sorry if it was not clear enough.

I would say this is fine since spark doesn't allow variable length arrays so if there are 2 rows with different dimensionality in foreign table, read will fail with or without this change.

What I can propose as a better way of fixing this is maybe querying the metadata first, and if it returns 0, we can fallback to querying array_ndim.

**Comment by yaooqinn** [comment]
The outcome of array_dims is data-dependent and volatile. I don't believe we can depend on it to create stable jobs. I don't think we can rely on it to create consistent jobs. Instead of experiencing occasional failures, I would prefer to fail at the outset.

**Comment by PetarVasiljevic-DB** [comment]
Maybe we're not on the same page here but this is what currently happens in Spark:
- If array column on Postgres has such values that dimension of these values across all rows is same then job will be stable.
- If array column on Postgres has such values that dimension of these values across all rows are not same, the job won't be stable.

With the proposed fix, behaviour stays the same:
- If array_ndims is same for all rows, job will be stable
- If array_ndims returns multiple values across all rows, job won't be stable.

Currently, jobs that are reading from CTAS tables will always fail. Have I misunderstood the term "stability of job" here?

**Comment by milastdbx** [comment]
If postgres array has row with different array dimensionality (array_dims returning different values for different rows) - then from Spark perspective there is really no correct answer to what `array_dims` we should use given that spark cannot process this rows anyway. Given that there is no correct answer, all answers are wrong answers, so returning `rand()` is same as reading it from metadata, as for Spark these are all incorrect.

**Comment by PetarVasiljevic-DB** [comment]
I see and it makes sense. Thanks @milastdbx. @yaooqinn may I suggest that we at least fallback to the value of 1 if we read 0 from metadata? It is expected to read the ArrayType so dimensionality of array should be at least 1. So something like this:

```
metadata.putLong("arrayDimension", Math.max(1, rs.getLong(1)))
```

It doesn't cover all the arrays still, but I am expecting that most of the users are using 1D arrays so this would cover most of the cases. Also, it is stable. 

And I see it as pure improvement not from the current code of PG dialect, but from the previous one, where we were reading all PG arrays (non-CTAS and CTAS) as 1D. With this code, we would support arrays (non-CTAS) of higher dimensions, but for the CTAS tables we would support only 1D arrays.

**Comment by yaooqinn** [comment]
It makes sense to me w/ 1 as the default dimension value or maybe we can fail directly when encountering 0. Also cc @cloud-fan

**Review by cloud-fan — APPROVED** [review]

**Comment by PetarVasiljevic-DB** [comment]
Can we merge? Test failures don't seem related to this RP.

**Comment by yaooqinn** [comment]
Can you rebase master and retry the GA?

**Comment by PetarVasiljevic-DB** [comment]
Tests are passing now.

**Review by yaooqinn — APPROVED** [review]

**Comment by yaooqinn** [comment]
Merged to master, thank you @PetarVasiljevic-DB
